### DISCUSSION TITLE
How can UEFI block downgradind the dbx?

### DISCUSSION THREAD

**Discussion by korpe1** [fwupd/fwupd#6890] [General]
How can UEFI block downgrading the dbx? Or is it UEFI?

dbx is located on writable storage, so how is this possible?

**Comment by superm1** [comment]
Dbx operations are only append. If you want to remove them, you'll need to reset the bios.

**Comment by hughsie** [comment]
In general you can't -- by design. Otherwise malware could just downgrade the dbx to a lower version and then install itself as a persistent firmware exploit. If it helps, sometimes resetting the BIOS settings to "factory defaults" clears the dbx completely.

**Reply by korpe1** [reply]
I understand the reason for this design, but I don't know how?

BootROM is the only physically immutable part of any firmware of computers (ofc you _can_ put whole firmware to ROM, but it disable ability to update). AFAIK mostly of PC BootROMs doesn't implement checking signatures of next steps; BootROM be in programming mode and flash or just download the actual firmware. So AFAIK everything in UEFI can be downgraded if you have image of old version. What prevents fwupd flash old version of dbx?

Or is this only "soft" restriction of the UEFI API? I assume that fwupd use something standard UEFI API to update. Is the case that this API simply doesn't accept downgrade the dbx? If so, I actually can downgrade with something lower level mechanism (programming mode of bootROM maybe?)?

**Reply by superm1** [reply]
https://blog.hansenpartnership.com/the-meaning-of-all-the-uefi-keys/#:~:text=All%20of%20the%20keys%20are%20stored%20in%20UEFI,of%20the%20key%20used%20to%20create%20the%20variable.:


> All of the keys are stored in UEFI secure variables (see section 7.2 of the UEFI specification).  A secure variable cannot be updated unless the person attempting the update can prove (with a digital signature on a specified payload, called the authentication descriptor) that they possess the private part of the key used to create the variable.  The UEFI specifications only allow two types of signing keys: X509 and RSA2048.  A secure variable may be cleared by writing an empty update (which must still contain a valid authentication descriptor).

**Reply by korpe1** [reply]
I appreciate your reply, but that doesn't answer my questions. It just tells how the key store perform and not how it's technically reached.

**Reply by superm1** [reply]
It does answer it though?  The UEFI API enforces the digital signature on the DBX and it completely abstracts the security of the access to the keystore.

**Reply by korpe1** [reply]
Okay, that's right. UEFI standard just sets abstract requirements for this. Anyway, to be clear, I wondering in real life examples of how this works?

If the UEFI doesn't allow downgrade the dbx without correct signature, what prevents me set bootROM to programming mode and downgrade the dbx? Or even directly write to the flash?

**Reply by superm1** [reply]
SPI write protect typically. If SPI write protect isn't set then yes wild west.

**Reply by korpe1** [reply]
> SPI write protect typically. If SPI write protect isn't set then yes wild west.

Can you be a little more detailed? SPI (Serial Peripheral Interface) doesn't tell much.

Do you mean that the SPI of bootROM is write protected? Really? The reason for existing bootROM is that if the firmware on flash ever fails we have something physically immutable backup plan that allows rewrite the whole flash. I think usually the bootROM programming mode is unrestricted.

**Reply by superm1** [reply]
How it works is different for Intel and AMD, but essentially there is a write protect pin that is controlled by a trusted entity.  OS can't change it, only the SoC can.  Think about it as going through different trust boundaries.  When the OS calls the UEFI API it will use an abstracted method to verify the signature.  UEFI API then works with the SoC to actually _arbitrate_ a write.

**Reply by korpe1** [reply]
> How it works is different for Intel and AMD, but essentially there is a write protect pin that is controlled by a trusted entity. OS can't change it, only the SoC can. Think about it as going through different trust boundaries. When the OS calls the UEFI API it will use an abstracted method to verify the signature. UEFI API then works with the SoC to actually arbitrate a write.

You didn't use reply, so please clarify, what part of my comment you talking about? Do you talk about the interface of bootROM or UEFI?

Also note that chips of Intel and AMD are mainly just CPUs and not SoCs. You might see integrated GPUs in laptop versions, but nothing level of SoCs used in mobile devices.

**Reply by hughsie** [reply]
> Also note that chips of Intel and AMD are mainly just CPUs and not SoCs

Mario works for AMD.

**Reply by korpe1** [reply]
>>Also note that chips of Intel and AMD are mainly just CPUs and not SoCs
>
>Mario works for AMD.

I think, also products by AMD are mainly CPU rather than SoCs. But this isn't important, just a sidenote.

**Reply by korpe1** [reply]
I really do appreciate both of your answers, but this conversation has going to sideways. I think maybe @superm1 think these things from different perspective than me. It isn't bad at all, but could it be better if something other try answer as well? Maybe @hughsie?

A set of  straight questions to clarify this conversation: 

 - BootROMs on PCs typically doesn't check the signature of firmware on flash, yes or no?

 - BootROMs typically gives ability to set unrestricted programming mode (alongside loading mode) and rewrite the flash, yes or no?

 - UEFI standard a spesify abstract requiment to UEFIs to blocking downgrade the dbx, yes or no?

 - UEFIs just simply doesn't allow downgrade the dbx through its APIs as per standard, yes or no?

 - Though UEFIs doesn't allow this, you can downgrade dbx with rewriting the flash via low-level interface of the bootROM (SPI in test pins etc)?

 - dbx is stored in [a secure cryptoprocessor](https://en.wikipedia.org/wiki/Secure_cryptoprocessor) that guard all uses of the dbx. So even bootROM or anything doesn't can downgrade the dbx without right authorization? Yes or no?

- Most of UEFIs although give ability to reset dbx to default in the UEFI setup manager, yes or no?

 - Many of UEFIs give ability to replace/set users own KEK (and PK, DB, DBX), how UEFI performs in this situation? What's the point of block DBX downgrades if the user can anyway replace KEK along with PK, DB, DBX?

Thank you already for all of your work, answers, and all other!

**Reply by superm1** [reply]
> I think, also products by AMD are mainly CPU rather than SoCs. But this isn't important, just a sidenote.

I respectfully disagree.  Legacy AMD products were true CPUs.  Modern products by AMD have a ton of non-X86 blocks that do other tasks.  NPU, ASP (TEE environment), GPU, MP1 (power management controller), MP2 (SFH), FCH.  There is a fabric for communication between blocks.  It's a true system on chip.  When I compare to a mobile SoC, the only significant architectural difference is that there is WWAN or WLAN built into the chip.

> I think maybe @superm1 think these things from different perspective than me. 

Surely I do.  But I'll try to comment from an outsider view.

> BootROMs on PCs typically doesn't check the signature of firmware on flash, yes or no?

For generic PCs:

- Intel Boot Guard will verify the signature.  
- For AMD, the ASP (formerly known as PSP) will verify the signature.

Some OEMs have implementations that the EC plays a role in trust for the SPI and SPI arbitration, but this can't be generically said.
For Chromebooks, Titan plays a role here.

> BootROMs typically gives ability to set unrestricted programming mode (alongside loading mode) and rewrite the flash, yes or no?

Not in most modern PC architecture.

> UEFI standard a spesify abstract requiment to UEFIs to blocking downgrade the dbx, yes or no?

Yes

> UEFIs just simply doesn't allow downgrade the dbx through its APIs as per standard, yes or no?

The DBX API works off of signed payloads, which are delivered as append operations.  I don't know if there is a way to distribute a signed variable that is a remove operation.  Might need to ask someone else on that unless Richard knows.

> Though UEFIs doesn't allow this, you can downgrade dbx with rewriting the flash via low-level interface of the bootROM (SPI in test pins etc)?

Yeah; if the implementation stores it a FV on the SPI and you end up turning off the machine and use a dediprog you can probably do this.  Whether or not the BIOS rejects it on the next boot is another story.

> dbx is stored in [a secure cryptoprocessor](https://en.wikipedia.org/wiki/Secure_cryptoprocessor) that guard all uses of the dbx. So even bootROM or anything doesn't can downgrade the dbx without right authorization? Yes or no?

Where it's stored is implementation vendor specific.

> Most of UEFIs although give ability to reset dbx to default in the UEFI setup manager, yes or no?

Yes

> Many of UEFIs give ability to replace/set users own KEK (and PK, DB, DBX), how UEFI performs in this situation? What's the point of block DBX downgrades if the user can anyway replace KEK along with PK, DB, DBX?

DBX operations can happen without a user present.  Think of it as if software can do it malware can do it too.
Replacing KEK/PK/DB/DBX require you to interactively open BIOS setup and do things.  Software can't directly do this.
